Add OpNovice GDML optical geometry and integration test

Rashika-Gupta opened the pull request:
This PR adds a new integration test for the `OpNovice` geometry with full optical properties exported from Geant4 as GDML implementing the following #1945 in the following way : 
1. Added `op-novice.gdml`.
2. Implemented `OpNoviceIntegrationMixin` to test optical physics - photon generation.
3. Added a new `OpNovice` test in `TrackingManagerIntegration`.
4. Updated `CMakeLists.txt` to build and run the new test.

Rashika-Gupta committed:
Added Optical Novice gdml

Rashika-Gupta committed:
Add OpNoviceIntegrationMinxin enabling optical physics

Rashika-Gupta committed:
Add OpNovice test

Rashika-Gupta committed:
Add Optical Novice Integration

Rashika-Gupta committed:
Setup test to run opnovice

Rashika-Gupta committed:
Merge remote-tracking branch 'upstream/develop' into add-opnovice-gdml

sethrj commented on test/accel/IntegrationTestBase.cc:
```diff
@@ -497,6 +497,50 @@ auto TestEm3IntegrationMixin::make_sens_det(std::string const& sd_name)
+{
+    PhysicsInput result = Base::make_physics_input();
+    result.em_bins_per_decade = 5;
+    return result;
```
I think we want OpNovice to *always* have optical physics, so you'll need to copy/adapt the code from `LarSphereOptical::make_physics_input` (from `TrackingManagerIntegration`) here and in `make_setup_options`.

sethrj commented on test/accel/CMakeLists.txt:
```diff
@@ -79,6 +79,8 @@ foreach(_basename IN ITEMS UserActionIntegration TrackingManagerIntegration)
     OFFLOAD cpu gpu
     RMTYPE serial
   )
+  # Test OpNovice
+  celeritas_add_integration_tests(${_basename} OpNovice run)
```
This line adds a bunch of different combinations, which is useful for testing primary functionality (GPU compatibility, thread safety) but not necessary for running the same type of problem that's already got a test. Replace with:

```suggestion
  # Test OpNovice
  celeritas_add_integration_tests(
    ${_basename} OpNoviceOptical run
    OFFLOAD cpu gpu g4
    RMTYPE ${_optical_rm_type}
  )
```
after moving the `_optical_rm_type` code above this loop.

sethrj commented on test/accel/CMakeLists.txt:
```diff
@@ -100,5 +107,11 @@ celeritas_add_integration_tests(
+  TrackingManagerIntegration OpNoviceOptical run
+  OFFLOAD cpu gpu g4
+  RMTYPE ${_optical_rm_type}
+)
```
Since the `OpNoviceOptical.run` tests are being added in the loop above, this is redundant.

sethrj commented on test/accel/TrackingManagerIntegration.test.cc:
```diff
@@ -409,6 +415,273 @@ TEST_F(LarSphereOptical, run)
+//---------------------------------------------------------------------------//
+class OpNovice : public OpNoviceIntegrationMixin, public TMITestBase
+{
+    void BeginOfEventAction(G4Event const* event) override
+    {
+        if (event->GetEventID() == 1)
+        {
+            for (auto i : range(event->GetNumberOfPrimaryVertex()))
+            {
+                G4PrimaryVertex* vtx = event->GetPrimaryVertex(i);
+                for (auto j : range(vtx->GetNumberOfParticle()))
+                {
+                    G4PrimaryParticle* p = vtx->GetPrimary(j);
+                    p->SetWeight(10.0);
+                }
+            }
+        }
+    }
```
There's no need to adjust the weight of the particles; this was done once to test that weight gets correctly sent to the hit objects.

sethrj reviewed (commented):
I hope I'm not reviewing this too early (I saw the review request but didn't notice it's still draft). It looks like there's a lot of unnecessarily duplicated code in the OpNovice test. Keep in mind:
1. We only want to test the problem with optical physics (unlike the LarSphere problem which I used to test some detector mechanics).
2. Eliminate tests/code that are the same as the other tests. For example, we already tested that the weights are being set correctly for EM particles, so there's no need to change the primary weights. (We *might* want to do something similar if we wanted to test that the weights are getting sent from the EM loop to the optical loop.) We don't need to test the UI commands unless it's something new.
3. Remember the primary purpose of the OpNovice test is to make sure that the physics is being set up and built correctly, so we need to find a way to check the constructed Celeritas physics and setup, and make sure that optical photons are being run.

Rashika-Gupta committed:
Address comments: Remove the repeated test cases

Rashika-Gupta committed:
Keep only optical physics and edit the primary event generator

Rashika-Gupta commented:
Addressed the comments. 
Also, increased the number of events to reproduce the failure at
`precondition failed: wavelength > 0`, which matches the earlier error I encountered.
```cpp
G4WT1 > [2/2]/forked/celeritas/src/accel/LocalTransporter.cc:386: Executing the first Celeritas stepping loop
/forked/celeritas/test/accel/IntegrationTestBase.cc:123: Failure
Failed
GeantExceptionHandler caught runtime error (1,celer.event.flush): from forked/celeritas/src/celeritas/optical/detail/OpticalUtils.hh:27: precondition failed: wavelength > 0
```

Rashika-Gupta commented:
> Addressed the comments. Also, increased the number of events to reproduce the failure at `precondition failed: wavelength > 0`, which matches the earlier error I encountered.
> 
> ```c++
> G4WT1 > [2/2]/forked/celeritas/src/accel/LocalTransporter.cc:386: Executing the first Celeritas stepping loop
> /forked/celeritas/test/accel/IntegrationTestBase.cc:123: Failure
> Failed
> GeantExceptionHandler caught runtime error (1,celer.event.flush): from forked/celeritas/src/celeritas/optical/detail/OpticalUtils.hh:27: precondition failed: wavelength > 0
> ```

This  occurs in  `wavelength_to_energy()` called from [`ScintillationGenerator.hh`](https://github.com/celeritas-project/celeritas/blob/b28f327bc9fb76c48250d13a2e9337ad641575e9/src/celeritas/optical/gen/ScintillationGenerator.hh#L162C3-L167C71)
```cpp
 sample_lambda_
        = NormalDistribution{component.lambda_mean, component.lambda_sigma};
    ExponentialDist sample_time(real_type{1} / component.fall_time);

    TrackInitializer photon;
    photon.energy = detail::wavelength_to_energy(sample_lambda_(rng));

```
For the specific instance
`(celeritas::NormalDistribution<>::result_type) wavelength = -3.8856919072923377E-9` 
`(celeritas::real_type) component.lambda_mean = 3.0849999999999999E-7` and `(celeritas::real_type) component.lambda_sigma = 8.9263721210270798E-8`

A possible fix could be 
```cpp
auto wavelength = sample_lambda_(rng);
if (wavelength <= 0) {
    CELER_LOG(warning) << "Invalid wavelength " << wavelength
                             << " — clamping to mean";
    wavelength = component.lambda_mean;
}
```

sethrj commented:
Thanks @Rashika-Gupta . Instead of warning and replacing with mean (i.e., lumping all of the PDF that's <= 0 into a delta function) I think it'd be better to resample so that we're truncating and rescaling the normal PDF:
```c++
real_type wavelength;
do {
wavelength = sample_lambda_(rng);
} while (wavelength <= 0);
```

Of course, the *correct* solution is to sample from the user PDF not the gaussian....

Rashika-Gupta commented:
> Thanks @Rashika-Gupta . Instead of warning and replacing with mean (i.e., lumping all of the PDF that's <= 0 into a delta function) I think it'd be better to resample so that we're truncating and rescaling the normal PDF:
> 
> ```c++
> real_type wavelength;
> do {
> wavelength = sample_lambda_(rng);
> } while (wavelength <= 0);
> ```
> 
> Of course, the _correct_ solution is to sample from the user PDF not the gaussian....

yes, would have preferred doing this but, did not want to mess with the existing code-base.

Rashika-Gupta committed:
Resample wavelength if less than zero

Rashika-Gupta committed:
Changing event number to 10

sethrj commented:
"Messing with" (i.e. "fixing") the code base is what these integration-level tests are supposed to drive us toward 😄 The code that existed before you isn't sacred, so it's always OK to bring up changes you think may be needed to get a problem to run.

sethrj committed:
Merge branch 'develop' into add-opnovice-gdml

Rashika-Gupta committed:
Fix direction of primary generator

sethrj commented on src/celeritas/optical/gen/ScintillationGenerator.hh:
```diff
@@ -161,10 +161,16 @@ CELER_FUNCTION TrackInitializer ScintillationGenerator::operator()(Generator& rn
+    do
+    {
+        wavelength = sample_lambda_(rng);
+    } while (CELER_UNLIKELY(wavelength <= 0));
```
It would be good to add a comment why this do loop is here (rejecting in the case of very large sigma and/or very small lambda)... and also to add an assertion after we get the component that `component.lambda_mean` is positive

sethrj commented on test/accel/UserActionIntegration.test.cc:
```diff
@@ -107,6 +107,26 @@ TEST_F(TestEm3, run)
+    rm.BeamOn(3);
+    cout << "initial run done" << endl;
+    rm.BeamOn(1);
+    cout << "second run done" << endl;
```
Most of our tests should probably have just a single run to make things quick: we want to make sure at least one of the tests has multiple runs, but I think that's covered.

Rashika-Gupta replied:
Originally, when the number of events was set to 1 or 2, the issue of generating photons with wavelength < 0 didn’t appear, likely due to limited statistics. Since that problem is now resolved, it should be fine to reduce the number of events again.

sethrj commented on test/accel/IntegrationTestBase.cc:
```diff
@@ -497,6 +497,76 @@ auto TestEm3IntegrationMixin::make_sens_det(std::string const& sd_name)
+    // Disable WLS (reemission not yet supported)
+    using WLSO = WavelengthShiftingOptions;
+    optical.wavelength_shifting = WLSO::deactivated();
+    optical.wavelength_shifting2 = WLSO::deactivated();
```
Do we need to disable WLS? I don't see any related data in the GDML input, so shouldn't Celeritas just ignore the models?

amandalund replied:
Probably not, though doesn't hurt to be explicit. If we used a GDML input with the data, would that enable it even though it's not fully supported?

sethrj replied:
Leaving it enabled is closer to the behavior of a user standalone application. @Rashika-Gupta added code that *should* disable the corresponding Celeritas optical physics if no materials have MFPs for it, so leaving it enabled is also a good check of that I think.

Rashika-Gupta replied:
Yes, I agree with you Seth that section of code (commit fb1fcf529 from PR #1933 ) was added specifically to handle cases where WLS MFPs were being created unintentionally in OpNovice example.

amandalund replied:
That's fair, so the OpNovice example creates the WLS process but then doesn't have any data for it?  I guess unlike other physics Geant4 might construct all the optical processes but then only "use" the ones that there's data for?

sethrj replied:
That is my understanding yes. I think Geant4's optical physics constructor builds all the processes but they're just no-ops for the most part.

sethrj reviewed (approved):
Great work!

sethrj committed:
Merge branch 'develop' into add-opnovice-gdml

Rashika-Gupta committed:
Addressed comments
